src/ray/object_manager/plasma/get_request_queue.h

```diff
@@ -0,0 +1,147 @@
+
+ private:
+  /// The timer that will time out and cause this wait to return to
+  /// the client if it hasn't already returned.
+  boost::asio::steady_timer timer_;
+  /// Whether or not if this get request is removed.
+  /// Once the get request is removed, any operation on top of the get request shouldn't
+  /// happen.
+  bool is_removed_ = false;
+};
+
+class GetRequestQueue {
+ public:
+  GetRequestQueue(instrumented_io_context &io_context,
+                  IObjectLifecycleManager &object_lifecycle_mgr)
+      : io_context_(io_context), object_lifecycle_mgr_(object_lifecycle_mgr) {}
+
+  /// Add a get request to get request queue. Note this will call callback functions
+  /// directly if all objects has been satisfied, otherwise store the request
+  /// in queue.
+  /// \param client the client where the request comes from.
+  /// \param object_ids the object ids to get.
+  /// \param timeout_ms timeout in millisecond, -1 is used to indicate that no timer
+  /// should be set. \param is_from_worker whether the get request from a worker or not.
+  /// \param object_callback the callback function called once any object has been
+  /// satisfied. \param all_objects_callback the callback function called when all objects
+  /// has been satisfied.
+  void AddRequest(const std::shared_ptr<ClientInterface> &client,
+                  const std::vector<ObjectID> &object_ids, int64_t timeout_ms,
+                  bool is_from_worker, ObjectReadyCallback object_callback,
+                  AllObjectReadyCallback all_objects_callback);
+
+  /// Remove all of the GetRequests for a given client.
+  ///
+  /// \param client The client whose GetRequests should be removed.
+  void RemoveGetRequestsForClient(const std::shared_ptr<ClientInterface> &client);
+
+  /// Handle a sealed object, should be called when an object sealed. Mark
+  /// the object satisfied and call object callbacks.
+  /// \param object_id the object_id to mark.
+  void MarkObjectSealed(const ObjectID &object_id);
+
+ private:
+  /// Remove a GetRequest and clean up the relevant data structures.
+  ///
+  /// \param get_request The GetRequest to remove.
+  void RemoveGetRequest(const std::shared_ptr<GetRequest> &get_request)
+      EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
+  /// Only for tests.
+  bool IsGetRequestExist(const ObjectID &object_id) const;
+
+  /// Called when objects satisfied. Call get request callback function and
+  /// remove get request in queue.
+  /// \param get_request the get request to be completed.
+  void OnGetRequestCompleted(const std::shared_ptr<GetRequest> &get_request);
+
+  instrumented_io_context &io_context_;
+
+  /// A hash table mapping object IDs to a vector of the get requests that are
+  /// waiting for the object to arrive.
+  absl::flat_hash_map<ObjectID, std::vector<std::shared_ptr<GetRequest>>>
+      object_get_requests_ GUARDED_BY(mu_);
+
+  /// Mutex to protect the `object_get_requests_`.
+  absl::Mutex mu_;
```

Comment:
yeah, tsan report data race for get_request_queue_test.cc because there is a single io_context to handle get timeout. We should find a solution to mock that, and the mutex should be revert.